[RDY] Job control fixes and pseudo terminal support

tarruda opened the pull request:
The end goal is to spawn bang commands connected to pseudo terminals, which will fix issues such as #1716. For windows an alternative implementation of pty_process.c will be required, probably on top of [winpty](https://github.com/rprichard/winpty/blob/master/include/winpty.h)

Pyrohh commented on src/nvim/os/pipe_process.c:
```diff
@@ -0,0 +1,111 @@
+} UvProcess;
+
+void pipe_process_init(Job *job, char **argv, bool enable_stdin,
+    bool enable_stdout, bool enable_stderr)
```
These args could be vertically aligned to help readability. I usually end up doing that if they don't all fit on one line.

Pyrohh commented on src/nvim/os/job_private.h:
```diff
@@ -0,0 +1,80 @@
+extern size_t stop_requests;
+extern uv_timer_t job_stop_timer;
+
+static inline void job_exit_callback(Job *job)
```
Does MSVC support this? IIRC they use __inline, but I'm not sure.

tarruda replied:
no clue here. cc @equalsraf

equalsraf replied:
The inline? We just typedef inline as __inline.

tarruda committed:
eval: Fix buffering of data in job autocommands

Job autocommands will no longer buffer data chunks that don't end in newlines
characters.

tarruda committed:
job: Send SIGTERM before calling job_stop in job_teardown

Send sigterm immediately since it can be caught by processes. If they don't
respond and are still alive after a while, SIGKILL will be sent.

tarruda committed:
job: Refactor process spawning and startup arguments

- process spawning was decoupled from the rest of the job control logic.  The
  goal is reusing it for spawning processes connected to pseudo terminal file
  descriptors.
- job_start now receives a JobOptions structure containing all the startup
  options.

tarruda committed:
job: Allow spawning jobs connected to pseudo terminals

Pyrohh commented:
My build is failing with this while linking `tty-test`:
```
[224/224] Linking C executable bin/tty-test
FAILED: : && /usr/lib/hardening-wrapper/bin/cc  -Wconversion -g  -Wl,--no-undefined test/functional/job/CMakeFiles/tty-test.dir/tty-test.c.o  -o bin/tty-test  -rdynamic ../.deps/usr/lib/libuv.a -ldl -lnsl -lrt && :
/usr/bin/ld: ../.deps/usr/lib/libuv.a(libuv_la-thread.o): undefined reference to symbol 'pthread_rwlock_trywrlock@@GLIBC_2.2.5'
/usr/lib/libpthread.so.0: error adding symbols: DSO missing from command line
collect2: error: ld returned 1 exit status
[224/224] Linking C executable bin/nvim
ninja: build stopped: subcommand failed.
Makefile:52: recipe for target 'nvim' failed
make: *** [nvim] Error 1
```

I get the error with or without `hardening-wrapper`, and this is with a clean build. It might have something to do with the fact that I'm on Arch Linux, which uses glibc 2.21

tarruda commented:
@Pyrohh strange that you would get this error now, all I did was add a custom executable target for testing the pty integration. What happens if you apply this patch:

```diff
diff --git a/CMakeLists.txt b/CMakeLists.txt
index caf345d..19ed0e7 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -276,7 +276,7 @@ get_compile_flags(NVIM_VERSION_CFLAGS)

 add_subdirectory(test/includes)
 add_subdirectory(config)
-add_subdirectory(test/functional/job) # compile pty test program
+# add_subdirectory(test/functional/job) # compile pty test program


 # Setup some test-related bits.  We do this after going down the tree because we
```

fwalch commented:
Similar error happens in the [Ubuntu PPA build](https://launchpadlibrarian.net/198586026/buildlog_ubuntu-vivid-i386.neovim_0.0.0%2Bgit201502240125%2B6~ubuntu15.04.1_BUILDING.txt.gz):

```
make[5]: Leaving directory '/build/buildd/neovim-0.0.0+git201502240125+6~ubuntu15.04.1/build'
[100%] Built target translations
make[5]: Entering directory '/build/buildd/neovim-0.0.0+git201502240125+6~ubuntu15.04.1/build'
Scanning dependencies of target tty-test
make[5]: Leaving directory '/build/buildd/neovim-0.0.0+git201502240125+6~ubuntu15.04.1/build'
make[5]: Entering directory '/build/buildd/neovim-0.0.0+git201502240125+6~ubuntu15.04.1/build'
[100%] Building C object test/functional/job/CMakeFiles/tty-test.dir/tty-test.c.o
Linking C executable ../../../bin/tty-test
/usr/bin/ld: ../../../../.deps/usr/lib/libuv.a(libuv_la-thread.o): undefined reference to symbol 'pthread_create@@GLIBC_2.1'
/lib/i386-linux-gnu/libpthread.so.0: error adding symbols: DSO missing from command line
collect2: error: ld returned 1 exit status
test/functional/job/CMakeFiles/tty-test.dir/build.make:86: recipe for target 'bin/tty-test' failed
make[5]: *** [bin/tty-test] Error 1
make[5]: Leaving directory '/build/buildd/neovim-0.0.0+git201502240125+6~ubuntu15.04.1/build'
CMakeFiles/Makefile2:2807: recipe for target 'test/functional/job/CMakeFiles/tty-test.dir/all' failed
make[4]: *** [test/functional/job/CMakeFiles/tty-test.dir/all] Error 2
make[4]: Leaving directory '/build/buildd/neovim-0.0.0+git201502240125+6~ubuntu15.04.1/build'
Makefile:117: recipe for target 'all' failed
make[3]: *** [all] Error 2
make[3]: Leaving directory '/build/buildd/neovim-0.0.0+git201502240125+6~ubuntu15.04.1/build'
Makefile:52: recipe for target 'nvim' failed
make[2]: *** [nvim] Error 2
make[2]: Leaving directory '/build/buildd/neovim-0.0.0+git201502240125+6~ubuntu15.04.1'
dh_auto_build: make -j1 returned exit code 2
debian/rules:4: recipe for target 'override_dh_auto_build' failed
make[1]: *** [override_dh_auto_build] Error 2
make[1]: Leaving directory '/build/buildd/neovim-0.0.0+git201502240125+6~ubuntu15.04.1'
debian/rules:11: recipe for target 'build-arch' failed
make: *** [build-arch] Error 2
dpkg-buildpackage: error: debian/rules build-arch gave error exit status 2
```

tarruda commented:
> undefined reference to symbol 'pthread_create

This error usually happens when linking against libuv without also linking to pthread, which is normally handled by the pkgconfig(or else even the travis build would fail). I have no clue why this error was introduced in the AUR/PPA builds. cc @jszakmeister

ajpaulson commented:
Hi - I'm getting the same issue when doing a build (even a fresh build) on Debian Jessie.

According to the output, pthreads is found:

    -- Looking for pthread_create in pthread - found
    -- Found Threads: TRUE  

I can't find confirmation that it is even attempting to link aside from that.

    Linking C executable ../../../bin/tty-test
    /usr/bin/ld: ../../../../.deps/usr/lib/libuv.a(libuv_la-thread.o): undefined reference to symbol           'pthread_rwlock_trywrlock@@GLIBC_2.2.5'
    //lib/x86_64-linux-gnu/libpthread.so.0: error adding symbols: DSO missing from command line
    collect2: error: ld returned 1 exit status

fwalch commented:
The PPA build for 12.04 worked, by the way.. guess that explains why Travis doesn't complain.

14.04:
```
/usr/bin/ld: ../../../../.deps/usr/lib/libuv.a(libuv_la-thread.o): undefined reference to symbol 'pthread_rwlock_trywrlock@@GLIBC_2.2.5'
```

14.10 and 15.04:
```
/usr/bin/ld: ../../../../.deps/usr/lib/libuv.a(libuv_la-thread.o): undefined reference to symbol 'pthread_rwlock_trywrlock@@GLIBC_2.2.5'
```

tarruda commented:
@fwalch is the build error present before this PR? I ask because I can't see what I could have changed to affect libuv linking

fwalch commented:
I can't actually find the revision used to build the packages in any logs :-/ The last successful build was at 2015-02-22 19:55, the one after that failed (there's only one build per day).

There are two different errors for x86 and x64:

x86:
```
libuv.a(libuv_la-thread.o): undefined reference to symbol 'pthread_create@@GLIBC_2.1'
```

x64:
```
libuv.a(libuv_la-thread.o): undefined reference to symbol 'pthread_rwlock_trywrlock@@GLIBC_2.2.5'
```

ajpaulson commented:
After running `git bisect` and double checking manually - the first build breaking commit is:

    0b8d3cb

Which breaks with a different error and is fixed in the following commit.

The first commit that breaks with this `libuv undefined reference to symbol...` error is:

    d7e560e

Which is a refactoring commit.

jszakmeister commented:
@tarruda See #2053 and #2054.

ajpaulson commented:
Confirmed fixed on Debian 7

justinmk commented on src/nvim/os/pty_process.c:
```diff
@@ -0,0 +1,225 @@
+  }
+
+  if (job->opts.stderr_cb) {
+    uv_pipe_open(&ptyproc->proc_stderr, dup(master));
```
coverity thinks these 3 `dup()`s are leaks.

tarruda replied:
I forgot to close the original fd. It has been fixed in #2076: https://github.com/tarruda/neovim/blob/builtin-terminal-emulation/src/nvim/os/pty_process.c#L108-L122

oni-link replied:
Coverity complains about not closing the (three additional) handles created by `dup(master)`. Closing `master` won't close these handles.

tarruda replied:
They are closed by libuv when the stream is closed: https://github.com/libuv/libuv/blob/v1.x/src/unix/stream.c#L1563-L1568

oni-link commented on src/nvim/os/pty_process.c:
```diff
@@ -0,0 +1,225 @@
+  init_termios(&termios);
+  uv_disable_stdio_inheritance();
+
+  int pid = forkpty(&master, NULL, &termios, &ptyproc->winsize);
```
Return value of `forkpty()` is of type `pid_t`. Is this always an `int`?

tarruda replied:
it should be, but I will use the typedef
